File: src/core/relational/aggregation.zig
// ============================================================================
// File: aggregation.zig
// Description: Single-pass aggregations over column values.
// Author/Maintainer: TakyonDB Contributors
// License: MIT. See LICENSE for details.
// ============================================================================

const std = @import("std");

/// Aggregation functions.
pub const AggFn = enum(u8) {
    Count = 0,
    Sum = 1,
    Avg = 2,
    Min = 3,
    Max = 4,
};

/// Accumulator for streaming aggregation (Kahan-friendly for f64 later).
pub const Acc = struct {
    count: usize = 0,
    sum: f64 = 0,
    min: f64 = std.math.inf(f64),
    max: f64 = -std.math.inf(f64),

    pub fn add(self: *Acc, v: f64) void {
        self.count += 1;
        self.sum += v;
        if (v < self.min) self.min = v;
        if (v > self.max) self.max = v;
    }

    pub fn result(self: *const Acc, fn_kind: AggFn) f64 {
        return switch (fn_kind) {
            .Count => @floatFromInt(self.count),
            .Sum => self.sum,
            .Avg => if (self.count == 0) 0 else self.sum / @as(f64, @floatFromInt(self.count)),
            .Min => if (self.count == 0) 0 else self.min,
            .Max => if (self.count == 0) 0 else self.max,
        };
    }
};

test "aggregation accumulator" {
    var a = Acc{};
    a.add(10);
    a.add(20);
    try std.testing.expectEqual(@as(f64, 30), a.result(.Sum));
    try std.testing.expectEqual(@as(f64, 15), a.result(.Avg));
    try std.testing.expectEqual(@as(f64, 10), a.result(.Min));
    try std.testing.expectEqual(@as(f64, 20), a.result(.Max));
    try std.testing.expectEqual(@as(f64, 2), a.result(.Count));
}

File: src/sdk/client/relational/aggregation.ts
/**
 * ============================================================================
 * File: aggregation.ts
 * Description: Single-pass aggregations over decoded rows.
 * Author/Maintainer: TakyonDB Contributors
 * License: MIT. See LICENSE for details.
 * ============================================================================
 */

import { Row } from './codec';

export type AggFn = 'count' | 'sum' | 'avg' | 'min' | 'max';

/**
 * Computes an aggregation over rows for a numeric column (count ignores column).
 *
 * Single pass, no intermediate array: the previous version built
 * `rows.map(...).filter(...)` (two arrays per call) and then called
 * `Math.min(...vals)` / `Math.max(...vals)`, which Makefan onto the call
 * stack. That is both an allocation and a real hazard: a wide enough column
 * overflows the argument limit and throws a RangeError instead of returning a
 * number. Iterating keeps the result defined for any row count.
 */
export function aggregate(rows: Row[], fn: AggFn, column?: string): number {
  if (fn === 'count') return rows.length;
  if (!column) throw new Error(`${fn} requires a column`);

  let sum = 0;
  let count = 0;
  let min = 0;
  let max = 0;
  let seen = false;

  for (const row of rows) {
    const v = row[column] as unknown;
    if (typeof v !== 'number') continue;
    if (!seen) {
      // Seed on the first numeric value so min/max are correct for
      // all-negative and all-positive columns alike.
      min = v;
      max = v;
      seen = true;
    } else {
      if (v < min) min = v;
      if (v > max) max = v;
    }
    sum += v;
    count++;
  }

  if (count === 0) return 0;
  switch (fn) {
    case 'sum':
      return sum;
    case 'avg':
      return sum / count;
    case 'min':
      return min;
    case 'max':
      return max;
  }
}
